[FLINK-23255][test] Introduce JUnit 5 dependencies

PatrickRen opened the pull request:
## What is the purpose of the change

This pull request introduces JUnit 5 dependencies into Flink project so that developers can write JUnit 5 style test cases. 

## Brief change log

- Add ```junit-jupiter``` dependency in root pom
- Add ```junit-vintage-engine``` and ```junit-jupiter-migrationsupport``` dependency for supporting existing JUnit 4 test cases.


## Verifying this change

This change is already covered by all existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented on pom.xml:
```diff
@@ -1578,7 +1591,7 @@ under the License.
 				<groupId>org.apache.maven.plugins</groupId>
 				<artifactId>maven-surefire-plugin</artifactId>
-				<version>2.22.1</version>
+				<version>2.22.2</version>
```
Why is this being bumped?

PatrickRen replied:
Thanks for the review @zentol! 2.22.2 [fixed a bug](https://issues.apache.org/jira/browse/SUREFIRE-1614) related to junit-vintage-engine, so I bumped the version.

AHeise commented on flink-end-to-end-tests/flink-end-to-end-tests-common-kafka/pom.xml:
```diff
@@ -49,8 +49,15 @@ under the License.
+
+		<dependency>
+			<groupId>org.junit.jupiter</groupId>
+			<artifactId>junit-jupiter-migrationsupport</artifactId>
```
Do we even need this? Afaik this is for using JUnit4 rules in JUnit5 tests. 
I had expected instead to see the vintage runner.

Same comment for the other modules.

AHeise reviewed (commented):
I have verified that the test count is plausible, however from the AZP run, I could not assess if  JUnit5 is actually used. The easiest way would be to temporarily add a test that fails and look at the stack trace.

PatrickRen commented:
Thanks for the review @AHeise ! 

Sorry for misunderstanding the usage of ```junit-jupiter-migrationsupport```. I have removed them in the latest commit. 

About validating whether Jupiter engine is truly used, I tried to check the stack trace but found that the stack is different under different execution environments. For example Intellij IDEA has its own engine so I can't capture keyword "jupiter" in the stack trace. Instead I added a temp test class ```JUnitJupiterTest``` with JUnit5 unique features. I think this case can prove that Jupiter engine truly works in the project.

AHeise commented:
> Thanks for the review @AHeise !
> 
> Sorry for misunderstanding the usage of `junit-jupiter-migrationsupport`. I have removed them in the latest commit.
> 
> About validating whether Jupiter engine is truly used, I tried to check the stack trace but found that the stack is different under different execution environments. For example Intellij IDEA has its own engine so I can't capture keyword "jupiter" in the stack trace. Instead I added a temp test class `JUnitJupiterTest` with JUnit5 unique features. I think this case can prove that Jupiter engine truly works in the project.

Sounds good to me and is a neat idea. We can double-check if the test count is the same and your new test appears on AZP and then this PR is good to go.

PatrickRen commented:
@AHeise Thanks again for the review! I made a rebase to sync with the latest master. 

I manually ran CI twice in my own AZP project, one with the JUnit 5 dependencies[1] and another without[2]. Number of tests are almost identical to each other (59,186 vs 59,189) only with slight diffs. 

Missing tests with JUnit5 dependencies: 
- org.apache.flink.testutils.junit.RetryRuleTest$TestClassWithAnnotation
- org.apache.flink.testutils.junit.RetryRuleTest$TestClassWithAnnotationOnMethod
- org.apache.flink.testutils.junit.RetryRuleTest$TestClassWithTestExpectingRuntimeException
- org.apache.flink.testutils.junit.RetryRuleTest$TestClassWithoutAnnotation
- org.apache.flink.test.checkpointing.LocalRecoveryITCase$EventTimeWindowCheckpointingITCaseInstance
- org.apache.flink.api.common.typeutils.CompositeSerializerTest$CompositeSerializerTestInstance
- org.apache.flink.api.java.typeutils.runtime.AbstractGenericTypeComparatorTest$ComparatorTestInstance
- org.apache.flink.api.java.typeutils.runtime.EitherSerializerTest$EitherSerializerTestInstance
- org.apache.flink.api.java.typeutils.runtime.LegacyRowSerializerTest$RowSerializerTestInstance
- org.apache.flink.api.java.typeutils.runtime.RowSerializerTest$RowSerializerTestInstance

Interesting thing is that all these tests are annotated with ```@Ignore```. Not sure why these cases are run in JUnit4. 

Additional tests with JUnit5 dependencies:
- org.apache.flink.test.junit5.JUnitJupiterTest * 6 (Intentionally added for testing Jupiter engine)
- org.apache.flink.connector.jdbc.xa.JdbcXaSinkMigrationTest#testCommitFromSnapshot (Not sure why this case is not run in JUnit4)

In general, I think all test cases are covered with by vintage engine, and new Jupiter engine is validated by ```JUnitJupiterTest```.

[1] With JUnit 5: https://dev.azure.com/renqs/Apache%20Flink/_build/results?buildId=42&view=ms.vss-test-web.build-test-results-tab
[2] Without JUnit 5: https://dev.azure.com/renqs/Apache%20Flink/_build/results?buildId=43&view=ms.vss-test-web.build-test-results-tab

AHeise reviewed (approved):
LGTM.
I'm assuming that one additional test was ignored on master or whatever you compared with. 
Let me know if I should merge. It would be best to do it at a time when you can also check if it causes issues with subsequent builds.
